### ISSUE TITLE
EtcdLedgerManager doesn't support layout versioning

### ISSUE THREAD

**Issue by ivankelly** [apache/bookkeeper#1863]
So we can't set the max ledger metadata format version, among other things.

**Comment by sijie** [comment]
@codingwangqi can you take a look at this?

**Comment by codingwangqi** [comment]
@ivankelly @sijie I don't quite understand the question here. are you talking about LedgerLayout? or something else?

**Comment by ivankelly** [comment]
@codingwangqi yes LedgerLayout. When a client or bookie writes and reads from a ledger manager they need to know the format. There's two parts to this. They need to know the layout of tree which they are reading, and they need to know the format of the contents of the nodes of the tree. In etcd, this is all implicit for now. Initially it was the same in ZK. Problems occur when you want to make a change. Some clients/bookies will understand the change while others will not (assuming not all client upgrade at the same instance), which can cause the clients to break each other. 

This is how it was with ZK. We went from a flat layout, with each ledger getting a znode under /ledgers to a hierarchical, where the ledger ID was segmented and the ledger metadata being stored in /ledgers/<XX>/<YY>/LXXYY. If a client assumed flat and added a ledger under /ledgers/LXXYY, stuff like garbage collection on the bookie (using hierarchical) would miss it and data would be lost.

LedgerLayout avoids this situation. When a client connects, it sees the layout which the client has been using and continues to use that layout, or gives up if it doesn't understand it. 

Recently I've extended this to include a version for the metadata contained within the nodes, which is how I spotted it was missing from etcd. Is anyone using etcd in production for this?

**Comment by codingwangqi** [comment]
@ivankelly : 

There is a EtcdLayoutManager which is used for storing and retrieving LedgerLayout. However my understand is the ledger layout was used to handle different variants of zookeeper-based ledger managers (as what you said, from flat to hierarchical). 

However Etcd is a key-value system, instead of tree system. so we don't think we are going to change how we structure key/values in Etcd. Even we have to do so in future, we can add that later (treating layout doesn't exist as v0).

We are getting rid of all the baggages in zookeeper and start with a clean design for Etcd. 

> Is anyone using etcd in production for this?

We are using it in staging environment.

**Comment by ivankelly** [comment]
@codingwangqi 

> However Etcd is a key-value system, instead of tree system. so we don't think we are going to change how we structure key/values in Etcd. Even we have to do so in future, we can add that later (treating layout doesn't exist as v0).

It's happening now. If you want to be able to use binary metadata with etcd you'll need to manage this.
Once CURRENT_METADATA_FORMAT_VERSION is bumped (it will be in a few releases), clients will start to break, so we should deal with this now.

**Comment by codingwangqi** [comment]
@ivankelly 

I am a bit confused now. 

Per my understanding, layout version is controlling the layout of how metadata is structured (e.g. in flat manager, it is just a flat structure in zk; in hierarchical, it is a hierarchical structure). However if you are just changing ledger metadata format, you should just bump `metadataFormatVersion` in metadata itself. You shouldn't change the layout version. My feeling is that you are using the wrong field. 

/cc @sijie

**Comment by ivankelly** [comment]
We are using the metadataFormatVersion in the metadata itself to control the version of the individual pieces of metadata. The field in layout version is to control it for the cluster as a whole, and it's to set the "maximum" format that will be stored.

This is needed for scenarios where there are multiple versions of the software reading and writing the same ledger metadata. Lets say you have a writer that understands version 3, and a reader that only understands version 2. If there is no clusterwide maximum, then the writer will write metadata that the reader cannot read.

**Comment by ivankelly** [comment]
Once #1866 is in, I'll push a change to pin etcd to version 2, so that this doesn't surprise us in the future.

**Comment by sijie** [comment]
sorry I am also a bit confused now.

 why do we need this maximum version format stored in ledger layout? I think I am lost in the sequence of changes. Can you write a BP for the metadata changes (both #1858 and #1866)? I am not sure I am following the changes, and I felt I am missing some pieces here and there. The BP should cover - what are needed to be changed, what tools are needed, what is the upgrade story and what documentation is required.

**Comment by sijie** [comment]
@ivankelly : also it seems that we added new fields to the layout format without bump layout version I #1858. it doesn't sound right to me. (I missed that in reviewing #1858). so I would suggest we hold on any metadata related PRs, and coming up with a BP with the whole picture about metadata changes, and make sure the community is aligned with the approach before merging any subsequent PRs.

**Comment by ivankelly** [comment]
> why do we need this maximum version format stored in ledger layout?

Because if we don't clients which understand binary will start writing it by default. The alternative is to configure each client whether we want to use binary or not.

> Can you write a BP for the metadata changes

Sure. I was supposed to write one in july, but it fell to the side. This will likely delay 4.9.

> also it seems that we added new fields to the layout format without bump layout version I #1858. it doesn't sound right to me.

Doesn't need a bump as the new field doesnt affect whether old readers can read it.

**Comment by sijie** [comment]
Lets come up with a BP, as we anyway need documentation for this.

**Comment by sijie** [comment]
4.9.0 will anyway be blocked for documentation if you don't have any. because metadata format changes require tools and documentation ready, otherwise 4.9.0 can't not be released.

**Comment by ivankelly** [comment]
Starting to think a client configuration option would be better. Can't remember why I went with a cluster wide. Need to think about it.

**Comment by eolivelli** [comment]
@ivankelly we can have several different BK clients on the same cluster, even from different applications.
It is common in my case to have BK 4.7 and 4.8 clients on the same cluster.

You have to be sure that new clients won't break old clients, but we also want a way to make the cluster metadata format version to be upgraded,

A client side configuration will be very tricky to handle, it would be handled even at application level.

**Comment by ivankelly** [comment]
moving this to dev@ as an issue about etcd isn't the right forum
